source/content/tools/pytest-api-examples.mkd

```diff
@@ -0,0 +1,322 @@
+---
+"kind": "article"
+"date": "2021-01-04 00:00 GMT"
+"title": "Pytest API Examples"
+---
+
+
+I use pytest in most python projects, and I've had a feeling that I haven't been
+been using most of the features it provides, since I tend to only use
+`@pytest.mark` from the API. I spent some time reading through the pytest
+documentation and playing with some examples in a [pytest jupyter notebook][] to
+get more familiarity with what's possible. After a few hours of playing around
+with pytest, realised there pytest has much more functionality than I had been.
+Read below for some of the examples. Most of this article is a reconstitution of
+what's in the [pytest documentation][] for my own self-learning. In addition to
+the pytest documentation there's useful videos, and plugins listed on the
+[awesome pytest][] GitHub repository.
+
+[pytest jupyter notebook]: https://github.com/michaelbarton/jupyer-pytest-api-examples
+[pytest documentation]: https://docs.pytest.org/en/stable/example/index.html
+[awesome pytest]: https://github.com/augustogoulart/awesome-pytest
+
+## Pytest Fixtures
+
+Fixtures are a large part of the pytest API, and the part I was least familiar
+with. Fixtures are included in test function by adding them as parameters to the
+function. The pytest API comes with a few builtin fixtures: useful ones for
+temporary files are [tmp_path][] and [tmp_path_factory][] shown below.
+
+[tmp_path_factory]: https://docs.pytest.org/en/stable/tmpdir.html#tmp-path-factory-example
+
+
+~~~ python
+def test_with_tmp_path(tmp_path: pathlib.Path):
```

Comment:
Does this require importing something?

Reply:
No, shouldn't do. Pytest provides all the fixtures for free.

Reply:
This confused me a lot when using fixtures in the project I'm working on right now. How does `pytest` know where to find `tmp_path`??

Reply:
Here's the source notebook: https://github.com/michaelbarton/jupyer-pytest-api-examples/blob/main/src/pytest_api_examples.ipynb

I used `nbconvert` to convert that to a markdown blog post.

Reply:
> how does pytest know where to find tmp_path

I assume there must be some kind of name resolution, where if an arguments parameters match the name of the fixture, then pytest passes the result of the fixture.

Reply:
And `tmp_path` is built in fixture function